design/design-doc-composition-functions.md

```diff
@@ -347,7 +354,39 @@ message Resource {
   google.protobuf.Struct resource = 1;
 
   // The resource's connection details.
+  // 
+  // * Crossplane will set this field in a RunFunctionRequest to communicate the
+  //   the observed connection details of a composite or composed resource.
+  //
+  // * A Function should set this field in a RunFunctionResponse to indicate the
+  //   desired connection details of the composite resource.
+  //
+  // * A Function should not set this field in a RunFunctionResponse to indicate
+  //   the desired connection details of a composed resource. This will be
+  //   ignored.
   map<string, bytes> connection_details = 2;
+
+  // Ready indicates whether the resource should be considered ready.
+  // 
+  // * Crossplane will never set this field in a RunFunctionRequest.
+  //
+  // * A Function should set this field to READY_TRUE in a RunFunctionResponse
+  //   to indicate that a desired composed resource is ready.
+  //
+  // * A Function should not set this field in a RunFunctionResponse to indicate
+  //   that the desired composite resource is ready. This will be ignored.
+  Ready ready = 3;
```

Comment:
@phisco @ezgidemirel @turkenh I'm curious how you feel about this.

Specifically, do you think it's okay to reuse this one `Resource` type in a bunch of different contexts? We currently use the one type to represent:

* The observed XR
* The desired XR
* Observed composed resources
* Desired composed resources

This keeps the protobuf schema simple, but calls for large comments like this explaining when each field will be honored and when each field will be ignored.

The alternative would be to have explicit `ObservedCompositeResource`, `DesiredCompositeResource`, ObservedComposedResource`, etc types. This would make it really obvious in the schema which fields are allowed where, at the cost of a more complex schema (and more complex corresponding code).

Reply:
Would it be sufficient to differentiate `Composed` from `Composite`? I believe what is set when observed vs. desired is more obvious.

Reply:
@turkenh Yeah, possibly.

At this point though my bias is to merge this as-is. We can make the change before beta in a follow-up PR if we decide one is needed.